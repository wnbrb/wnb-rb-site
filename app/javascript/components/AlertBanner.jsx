import React from 'react';
import PropTypes from 'prop-types';
import 'stylesheets/alert-banner';

const AlertBanner = ({ message, type, onClose }) => {
    const emojis = { success: '👍🏼', error: '😨', warning: '⚠️❗️' };

    const handleAlertClose = () => {
        onClose();
    };

    return (
        <div className={`alert-banner ${type}`}>
            <p className="message m-0">
                <span className="emoji-icon">{emojis[type]}</span>
                <span className="ml-1">{message}</span>
            </p>
            <button className="close-button" onClick={handleAlertClose}>
                X
            </button>
        </div>
    );
};

AlertBanner.propTypes = {
    type: PropTypes.oneOf(['success', 'error', 'warning']),
    message: PropTypes.string,
    onClose: PropTypes.func,
};

export default AlertBanner;
