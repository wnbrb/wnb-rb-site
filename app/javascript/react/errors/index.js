export class UnauthorizedError extends Error {
    constructor(message) {
        super(message);
        this.name = 'UnauthorizedError';
    }
}
