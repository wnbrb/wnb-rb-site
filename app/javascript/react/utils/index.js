const MS_IN_DAY = 24 * 60 * 60 * 1000;

export const postedAtString = (createdAtString) => {
    const today = new Date();
    const postedDate = new Date(createdAtString);

    const diff = (+today - +postedDate) / MS_IN_DAY;

    if (diff <= 1) {
        return 'today';
    }

    if (diff < 7) {
        return `${Math.floor(diff)} days ago`;
    }

    if (diff < 30) {
        const weekDiff = Math.floor(diff / 7);
        const weekOrWeeks = weekDiff === 1 ? 'week' : 'weeks';

        return `${weekDiff} ${weekOrWeeks} ago`;
    }

    const monthDiff = Math.floor(diff / 30);
    const monthOrMonths = monthDiff === 1 ? 'month' : 'months';

    return `${monthDiff} ${monthOrMonths} ago`;
};
