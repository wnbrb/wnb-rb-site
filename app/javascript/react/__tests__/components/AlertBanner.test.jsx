import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import AlertBanner from '../../components/AlertBanner';

describe('AlertBanner Component', () => {
    const mockOnClose = jest.fn();

    beforeEach(() => {
        mockOnClose.mockClear();
    });

    // Snapshot tests
    it('should match snapshot with success type', () => {
        const { container } = render(
            <AlertBanner
                type="success"
                message="Operation completed successfully"
                onClose={mockOnClose}
            />,
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot with error type', () => {
        const { container } = render(
            <AlertBanner type="error" message="Something went wrong" onClose={mockOnClose} />,
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot with warning type', () => {
        const { container } = render(
            <AlertBanner type="warning" message="Please check your input" onClose={mockOnClose} />,
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    // Functional tests
    it('should render message correctly', () => {
        const message = 'Test alert message';
        const { getByText } = render(
            <AlertBanner type="success" message={message} onClose={mockOnClose} />,
        );
        expect(getByText(message)).toBeInTheDocument();
    });

    it('should display correct emoji for each type', () => {
        const { rerender, container } = render(
            <AlertBanner type="success" message="Success" onClose={mockOnClose} />,
        );
        expect(container.querySelector('.emoji-icon')).toHaveTextContent('👍🏼');

        rerender(<AlertBanner type="error" message="Error" onClose={mockOnClose} />);
        expect(container.querySelector('.emoji-icon')).toHaveTextContent('😨');

        rerender(<AlertBanner type="warning" message="Warning" onClose={mockOnClose} />);
        expect(container.querySelector('.emoji-icon')).toHaveTextContent('⚠️❗️');
    });

    it('should apply correct CSS class for type', () => {
        const { container } = render(
            <AlertBanner type="error" message="Error message" onClose={mockOnClose} />,
        );
        expect(container.firstChild).toHaveClass('alert-banner', 'error');
    });

    it('should call onClose when close button is clicked', () => {
        const { getByText } = render(
            <AlertBanner type="success" message="Test message" onClose={mockOnClose} />,
        );

        fireEvent.click(getByText('X'));
        expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    it('should have accessible close button', () => {
        const { container } = render(
            <AlertBanner type="success" message="Test message" onClose={mockOnClose} />,
        );

        const closeButton = container.querySelector('.close-button');
        expect(closeButton).toBeInTheDocument();
        expect(closeButton.tagName).toBe('BUTTON');
    });
});
