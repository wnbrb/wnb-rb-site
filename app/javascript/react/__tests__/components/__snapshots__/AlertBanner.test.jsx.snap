// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertBanner Component should match snapshot with error type 1`] = `
<div
  class="alert-banner error"
>
  <p
    class="message m-0"
  >
    <span
      class="emoji-icon"
    >
      😨
    </span>
    <span
      class="ml-1"
    >
      Something went wrong
    </span>
  </p>
  <button
    class="close-button"
  >
    X
  </button>
</div>
`;

exports[`AlertBanner Component should match snapshot with success type 1`] = `
<div
  class="alert-banner success"
>
  <p
    class="message m-0"
  >
    <span
      class="emoji-icon"
    >
      👍🏼
    </span>
    <span
      class="ml-1"
    >
      Operation completed successfully
    </span>
  </p>
  <button
    class="close-button"
  >
    X
  </button>
</div>
`;

exports[`AlertBanner Component should match snapshot with warning type 1`] = `
<div
  class="alert-banner warning"
>
  <p
    class="message m-0"
  >
    <span
      class="emoji-icon"
    >
      ⚠️❗️
    </span>
    <span
      class="ml-1"
    >
      Please check your input
    </span>
  </p>
  <button
    class="close-button"
  >
    X
  </button>
</div>
`;
